optuna/storages/_rdb/storage.py

```diff
@@ -863,6 +879,28 @@ def _build_frozen_trial_from_trial_model(self, trial: "models.TrialModel") -> Fr
         else:
             values = None
 
+        unsorted_intermediates: dict[int, dict[int, float]] = defaultdict(dict)
+        for v in trial.intermediate_values:
+            unsorted_intermediates[v.step][
+                v.intermediate_value_index
+            ] = v.stored_repr_to_intermediate_value(
+                v.intermediate_value, v.intermediate_value_type
+            )
+        intermediate_values: Dict[int, Union[float, Sequence[float]]] = {}
+        if unsorted_intermediates:
+            num_intermediate_values_at_same_step = len(unsorted_intermediates[v.step])
+            if num_intermediate_values_at_same_step != 1:
+                for step, intermediate_indices_and_values in unsorted_intermediates.items():
+                    intermediate_values[step] = tuple(
+                        intermediate_value
+                        for _, intermediate_value in sorted(
+                            intermediate_indices_and_values.items()
+                        )
+                    )
+            else:
+                for step, intermediate_indices_and_values in unsorted_intermediates.items():
+                    intermediate_values[step] = intermediate_indices_and_values[0]
+
```

Comment:
I'm not sure `intermediate_values` is sorted by `intermediate_value_index` in case, but if not, we can simplify this part.